text/0000-argument-lifetimes.md

```diff
@@ -0,0 +1,476 @@
+come from outside of its stack frame.
+
+The most typical case is producing new borrowed data from already-borrowed
+data. For example, consider a "getter" method:
+
+```rust
+struct MyStruct {
+    field1: Foo,
+    field2: Bar,
+}
+
+impl MyStruct {
+    fn get_field1(&self) -> &Foo {
+        &self.field1
+    }
+}
+```
+
+Here we're making what looks like a "fresh" borrow, it's "derived" from the
+existing borrow of `self`, and hence fine to return back to our caller; the
+actual `MyStruct` value must live outside our stack frame anyway.
+
+### Pinpointing borrows with lifetimes
+
+For Rust to guarantee safety, it needs to track the *lifetime* of each loan,
+which says *for what portion of code the loan is valid*.
+
+In particular, each `&` type also has an associated lifetime---but you can
+usually leave it off. The reason is that a lot of code works like the getter
+example above, where you're returning borrowed data which could only have come
+from the borrowed data you took in. Thus, in `get_field1` the lifetime for
+`&self` and for `&Foo` are assumed to be the same.
+
+Rust is conservative about leaving lifetimes off, though: if there's any
+ambiguity, you need to say explicitly state the relationships between the
+loans. So for example, the following function signature is *not* accepted:
+
+```rust
+fn select(data: &Data, params: &Params) -> &Item;
+```
+
+Rust cannot tell how long the resulting borrow of `Item` is valid for; it can't
+deduce its lifetime. Instead, you need to connect it to one or both of the input
+borrows:
+
+```rust
+fn select(data: &'data Data, params: &Params) -> &'data Item;
+fn select(data: &'both Data, params: &'both Params) -> &'both Item;
+```
+
+This notation lets you *name* the lifetime associated with a borrow and refer to
+it later:
+
+- In the first variant, we name the `Data` borrow lifetime `'data`, and make
+clear that the returned `Item` borrow is valid for the same lifetime.
+
+- In the second variant, we give *both* input lifetimes the *same* name `'both`,
+which is a way of asking the compiler to determine their "intersection"
+(i.e. the period for which both of the loans are active); we then say the
+returned `Item` borrow is valid for that period (which means it may incorporate
+data from both of the input borrows).
+
+In addition, when an argument has only one lifetime you could be referencing,
+like `&Data`, you can refer to that lifetime by the argument's name:
+
+```rust
+fn select(data: &Data, params: &Params) -> &'data Item;
+```
+
+Otherwise, you are not allowed to use an argument's name as a lifetime.
+
+## `struct`s and lifetimes
+
+Sometimes you need to build data types that contain borrowed data. Since those
+types can then be used in many contexts, you can't say in advance what the
+lifetime of those borrows will be. Instead, you must take it as a parameter:
+
+```rust
+struct VecIter<'vec, T> {
+    vec: &'vec Vec<T>,
```

Comment:
I assume because we don't want to write `VecIter<T>`, we want to write `VecIter<'_, T>`. It would be weirdly inconsistent to let you omit the lifetime parameters when defining `VecIter` despite requiring them at every use site.